tests/integration/class-s3-media-sync-test.php

```diff
@@ -0,0 +1,133 @@
+<?php
+/**
+ * Integration tests for S3_Media_Sync
+ *
+ * @package S3_Media_Sync
+ */
+
+namespace S3_Media_Sync\Tests\Integration;
+
+use S3_Media_Sync;
+
+/**
+ * S3_Media_Sync test case.
+ *
+ * @covers S3_Media_Sync
+ */
+class S3_Media_Sync_Test extends WP_Integration_Test_Case {
+
+	private array $complete_settings
+		= [
+			'bucket'     => 'test-bucket',
+			'key'        => 'test-key',
+			'secret'     => 'test-secret',
+			'region'     => 'test-region',
+			'object_acl' => 'public-read',
+		];
+
+	private array $incomplete_settings
+		= [
+			'bucket'     => '',
+			'key'        => '',
+			'secret'     => '',
+			'region'     => '',
+			'object_acl' => 'public-read',
+		];
+
+	/**
+	 * Helper method for that checking media sync hooks are registered.
+	 *
+	 * @param string $assertion The assertion to use.
+	 */
+	private function assert_media_syncs_hooks_registered( string $assertion ): void {
+		self::$assertion( has_filter( 'wp_handle_upload', [ $this->s3_media_sync, 'add_attachment_to_s3' ] ) );
+		self::$assertion( has_action( 'delete_attachment', [ $this->s3_media_sync, 'delete_attachment_from_s3' ] ) );
+		self::$assertion( has_filter( 'wp_save_image_editor_file', [ $this->s3_media_sync, 'add_updated_attachment_to_s3' ] ) );
+	}
+
+	/**
+	 * @testdox Allow media syncs when required settings set
+	 *
+	 * @throws \ReflectionException
+	 */
+	public function test_allow_media_syncs_when_required_settings_set(): void {
+		parent::set_private_property(
+			$this->s3_media_sync::class,
+			$this->s3_media_sync,
+			'settings',
+			$this->complete_settings
+		);
```

Comment:
See https://github.com/Automattic/s3-media-sync/pull/35#discussion_r1951696988 - in that, having to force change the value of a private class property with Reflection seems like a huge code smell (I know you're working with what you've got for this particular hotfix). Having a Settings object, injected into the main class constructor, means that the class under test could just be instantiated here with a data provider of arrays / array key values.

Reply:
Agreed. I'm going to attempt to flush this out in https://github.com/Automattic/s3-media-sync/pull/35.